src/TokenInterface.php

```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Dhii\Parser\Tokenizer;
+
+use Dhii\Data\Container\ContainerInterface;
+
+/**
+ * Represents a token.
+ *
+ * A token is a piece of data at a specific location. It has domain-specific
+ * data attached to it.
+ *
+ * @since [*next-version*]
+ */
+interface TokenInterface extends
```

Comment:
How will consumers identify tokens? By their class name? By using `get()`?

A token, in my opinion, has some code-like string type intrinsic to it that indicates what it represents. This is because _that_ is what is most needed by parsers in order to know context, perform lookahead, etc.

Reply:
Context is acquired via `ContainerInterface#get()`. However, it's true that the "type" of the token is what's most important. I'm not sure how I meant to identify tokens. Maybe it deserves a `getKey()`.

Reply:
Yes I think it deserves a `getKey()`.

Also, what do you mean by "context" in the context (pun) of a tokenizer? Do you mean data about a specific token instance, which would include the read input?

Reply:
@mecha, yes, exactly.

Reply:
Right. So there is no standard method for retrieving the image of a token? I would suggest adding such a method. It would give consumers a complete API for retrieving the key, image, line, column and other misc. data for a token.

Or ... perhaps make it stringable!

Reply:
Stringable IMHO is not suitable because a token represents more than a string, i.e. it's not enough to only know it's "content" in order to work with it. What about `getValue()`? That could compliment `getKey()` very well.

Reply:
I considered `getValue()` briefly while typing my previous message. IMO, it would have to be re-declared with a `string|Stringable` return type. Otherwise, just introduce a `getImage()`. I don't know what to choose to be honest. The former sounds more Dhii-like, the latter has a sense of familiarity since it uses CS terminology. But in reality, the value of a token is its image. So I'm equally fine with either.

Reply:
Not sure why the `Stringable` type. It would then have to be resolved. I would just declare `string`. Good?

Reply:
I mean I can't anticipate cases where you'd want `Stringable`, so I just mentioned it as an added flexibility. I don't think it adds any complexity with regards to resolving.

But just `string` is ok too. It just means that if the image is a `Stringable` instead of a `string`, it needs to be "rendered" by the token before being exposed, rather than rendered by the consumer. Given that an image shouldn't change depending on render time, it's effectively the same thing.

Reply:
Parsing is only done on strings per definition. Not sure why it would be anything else other than a string. And yes, the added complexity is that the returned value would have to be resolved first, such as with `_normalizeString()`. So, string it is. Addink.